src/ui/map.js

```diff
@@ -2167,7 +2184,50 @@ class Map extends Camera {
      * @returns {Object} terrain Terrain specification properties of the style.
      */
     getTerrain(): Terrain | null {
-        return this.style.getTerrain();
+        return this.style ? this.style.getTerrain() : null;
+    }
+
+    /**
+     * Sets the fog property of the style.
+     * @param fog The fog properties to set. Must conform the [Mapbox Style Specification](https://docs.mapbox.com/mapbox-gl-js/style-spec/root/#fog).
+     * If `null` or `undefined` is provided, this function call removes the fog from the map.
+     * @returns {Map} `this`
+     * @example
+     * map.setFog({
+     *  "range": [1.0, 12.0],
+     *  "color": 'white',
+     *  "horizon-blend": 0.1
+     * });
+     */
+    setFog(fog: FogSpecification) {
+        this._lazyInitEmptyStyle();
+        this.style.setFog(fog);
+        return this._update(true);
+    }
+
+    /**
+     * Returns the fog specification or `null` if fog is not set on the map.
+     *
+     * @returns {Object} fog Fog specification properties of the style.
+     */
+    getFog(): Fog | null {
+        return this.style ? this.style.getFog() : null;
+    }
+
+    /**
+     * Returns the fog opacity for a given geolocation.
+     *
+     * An opacity of 0 means that there is no fog contribution for the given geolocation
+     * while a fog opacity of 1.0 means the geolocation is fully obscured by the fog effect.
+     *
+     * If there is no fog set on the map, this function will simply return 0.
+     *
+     * @param {LngLatLike} lnglat The geographical location to evaluate the fog on.
+     * @returns {number} A value between 0 and 1 representing the fog opacity, where 1 means fully within, and 0 means not affected by the fog effect.
+     */
+    queryFogOpacity(lnglat: LngLatLike): number {
```

Comment:
Do we have clear use cases for this? If not, maybe we should keep this private for now.

At first I was thinking it might be useful for custom layers but I think we might need a bunch of other things to make custom layers work.

Reply:
It's used for markers and would be useful for similar geolocated drawables (think gl-draw, pucks, or anything that does not necessarily need a full and custom gl pipeline). I don't think this falls in the category of API that will mature more than what it is, but I'm also fine with naming it in the private scope for now.

Reply:
Addressed in https://github.com/mapbox/mapbox-gl-js/pull/10564/commits/e5abf7594f814468e622f6953806216c1c179389